src/CoordinatorClusterClock.cc

```diff
@@ -0,0 +1,168 @@
+/* Copyright (c) 2014-2015 Stanford University
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <string>
+
+#include "CoordinatorClusterClock.h"
+#include "CoordinatorClusterClock.pb.h"
+#include "ShortMacros.h"
+#include "Util.h"
+
+namespace RAMCloud {
+
+/**
+ * Constructor for the CoordinatorClusterClock.
+ *
+ * \param context
+ *      Overall information about the RAMCloud server and provides access to
+ *      externalStorage (externalStorage must be non-null).
+ */
+CoordinatorClusterClock::CoordinatorClusterClock(Context *context)
+    : mutex()
+    , startingSysTimeUs(Cycles::toMicroseconds(Cycles::rdtsc()))
+    , startingClusterTimeUs(recoverClusterTime(context->externalStorage))
+    , safeClusterTimeUs(startingClusterTimeUs)
+    , updater(context, this)
+{}
+
+/**
+ * Returns the current cluster time.  Repeated calls this this method are
+ * guaranteed to return monotonically increase cluster times, even across
```

Comment:
Does "monotonically increasing" imply that no two calls ever return the same value?

I'm wondering if you need to say "monotonically non-decreasing" instead.